### PR TITLE
OpenGL Integration

### PR DESCRIPTION
**Pull request by bschwind**
In our use case, we have an OpenGL texture which already contains the image we want to run a network on. It has already been resized/cropped/formatted to the kind of raw format darknet expects so we'd like to run it straight through darknet instead of downloading to a CPU buffer and passing it to darknet, which will then re-upload it to GPU memory.

This code has been tested in a real-world application and it works, while reducing contention on the PCIe bus. Happy to add feature flags or other adjustments, I wasn't sure how best to handle that.

BTW for anyone curious, here's a little shader to convert RGB(A) texture data to a format that Darknet expects (all red bytes, then all green bytes, then all blue bytes):

```glsl
#version 410

uniform sampler2D source;

in vec2 v_Uv;
out vec4 Target0;

void main() {
    vec2 uv = vec2(v_Uv.x, mod(v_Uv.y, 1.0 / 3.0) * 3.0);
    vec3 color = texture(source, uv).rgb;

    float is_red = float(v_Uv.y < (1.0 / 3.0));
    float is_green = float(v_Uv.y >= (1.0 / 3.0) && v_Uv.y < (2.0 / 3.0));
    float is_blue = float(v_Uv.y >= (2.0 / 3.0));

    float val = (color.r * is_red)
        + (color.g * is_green)
        + (color.b * is_blue);

    Target0 = vec4(val, 0.0, 0.0, 1.0);
}
```

### CODE REVIEW

File: include/darknet.h [hunk 1]
```diff
@@ -1032,6 +1032,7 @@ LIB_API void diounms_sort(detection *dets, int total, int classes, float thresh,
 // network.h
 LIB_API float *network_predict(network net, float *input);
 LIB_API float *network_predict_ptr(network *net, float *input);
+LIB_API float *network_predict_gl_texture(network *net, uint32_t texture_id);
```
Comment (bschwind) [inline include/darknet.h:1035]: Should this be behind an `#ifdef GPU` or similar gate?

File: src/network_kernels.cu [hunk 2]
```diff
@@ -747,3 +747,90 @@ float *network_predict_gpu(network net, float *input)
...
+{
+    if (net.gpu_index != cuda_get_device())
+        cuda_set_device(net.gpu_index);
+    int size = get_network_input_size(net) * net.batch;
+
+    // Map the OpenGL texture resource so CUDA can access it.
+    cudaGraphicsResource_t graphics_resource = NULL;
+    unsigned int flags = cudaGraphicsRegisterFlagsReadOnly;
+    CHECK_CUDA(cudaGraphicsGLRegisterImage(&graphics_resource, texture_id, GL_TEXTURE_2D, flags));
+    CHECK_CUDA(cudaGraphicsMapResources(1, &graphics_resource, 0));
```
Comment (bschwind) [inline src/network_kernels.cu:761]: While maybe not _ideal_ to register and unregister textures on every frame, the alternative was keeping some sort of map of texture IDs to resources, which felt more intrusive and increases complexity in this PR.

File: src/network_kernels.cu [hunk 3]
```diff
@@ -747,3 +747,90 @@ float *network_predict_gpu(network net, float *input)
...
+    }
+
+    float *out = get_network_output_gpu(net);
+    reset_wait_stream_events();
+    //cuda_free(state.input);   // will be freed in the free_network()
+
+    // Unmap the OpenGL texture.
+    cudaGraphicsUnmapResources(1, &graphics_resource, 0);
+    cudaGraphicsUnregisterResource(graphics_resource);
+
```
Comment (bschwind) [inline src/network_kernels.cu:834]: A lot of this code was copy-pasted from `network_predict_gpu`, I'm happy to de-duplicate it if desired.

**Comment by cenit** [comment]
Thanks for your PR
I will review it more in the coming days. In the meantime is it possible for you to work on errors highlighted by CI?

**Comment by bschwind** [comment]
Hi @cenit, certainly! I'll try to get the CI jobs cleaned up today, seems to just be some missing headers.

**Comment by cenit** [comment]
don't worry about the opencv3 regression, it's due to an upstream bug which will be fixed very soon (https://github.com/microsoft/vcpkg/pull/26378)

please consider errors in windows when CUDA is available. It looks like your modifications are not so windows-compatible

**Comment by bschwind** [comment]
@cenit I don't have a windows machine immediately available but I'll try to set one up soon to test it there. I assume it's also just a matter of installing opengl and/or pointing it to the correct headers and libraries to link.

**Comment by cenit** [comment]
might be just the matter of adding a "opengl" reference to vcpkg.json manifest then!

**Comment by cenit** [comment]
besides proper linking in the CMakeLists.txt then... why is it working without otherwise?

**Comment by bschwind** [comment]
> might be just the matter of adding a "opengl" reference to vcpkg.json manifest then!

I think this will help, yep. The question then becomes: Should I enable opengl as a dependency only when the cuda feature is enabled, or should I be adding a new feature flag for opengl specifically?

**Comment by cenit** [comment]
I'd add opengl separately, as a feature that depends on the cuda feature. In this way we might be able to properly cover many operating systems better. There might be corner cases where cuda is available but an opengl compatible layer is not

**Comment by cenit** [comment]
> don't worry about the opencv3 regression, it's due to an upstream bug which will be fixed very soon ([microsoft/vcpkg#26378](https://github.com/microsoft/vcpkg/pull/26378))
> 
> please consider errors in windows when CUDA is available. It looks like your modifications are not so windows-compatible

upstream issue has been resolved, so I retriggered CI for your PR, in order to remove spurious failures

**Comment by bschwind** [comment]
Thank you, @cenit . I should have access to a windows machine tomorrow so hopefully I'll get this all green then. I'm in Tokyo and I think we're unfortunately in quite different timezones so the back and forth takes a bit longer.

**Comment by bschwind** [comment]
@cenit could I actually get your help on the windows build? I pretty much never develop on windows and trying it out locally is frustrating, to understate it.

**Comment by cenit** [comment]
ok sure, just give me some time

**Comment by bschwind** [comment]
Thank you - in the meantime, I can still try to add the dependencies properly to CMakeLists and vcpkg.json. I don't know if "features depending on other features" is a first-class citizen in CMake, but it seems to exist as a concept in vcpkg.json.

**Comment by cenit** [comment]
> Thank you - in the meantime, I can still try to add the dependencies properly to CMakeLists and vcpkg.json. I don't know if "features depending on other features" is a first-class citizen in CMake, but it seems to exist as a concept in vcpkg.json.

https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html
cmake_dependent_option might be your friend in this case (might... because I have still not explored your feature)

**Comment by 1027663760** [comment]
Does it work on windows system?

**Comment by cenit** [comment]
not for now, and sorry for being so late helping PR author on it. It's an extremely busy period for me, sorry again

**Comment by bschwind** [comment]
No worries, @cenit ! There's no rush, I'm quite busy as well so we'll get to it when we can.

File: src/blas.c [hunk 4]
```diff
@@ -342,32 +342,32 @@ void fill_cpu(int N, float ALPHA, float *X, int INCX)
     }
 }
 
-void deinter_cpu(int NX, float *X, int NY, float *Y, int B, float *OUT)
+void deinter_cpu(int NX, float *X, int NY, float *Y, int B, float *OUTPUT)
```
Comment (bschwind) [inline src/blas.c:345]: `OUT` was conflicting with a macro defined in a Windows header, renamed to `OUTPUT` to avoid the conflict.

**Comment by bschwind** [comment]
@cenit (@1027663760 too) this _should_ work on Windows now, hopefully...

I sat down with my coworker (thank you @skywhale!) and got it building on a windows machine.

I also tried to get the functionality behind a more proper feature flag, let me know if I messed that up.

**Comment by 1027663760** [comment]
> @cenit (@1027663760 too) this _should_ work on Windows now, hopefully...
> 
> I sat down with my coworker (thank you @skywhale!) and got it building on a windows machine.
> 
> I also tried to get the functionality behind a more proper feature flag, let me know if I messed that up.

Set like this?
[net]
use_cuda_graph=1

**Comment by bschwind** [comment]
@1027663760 it's a CMake option, `ENABLE_CUDA_OPENGL_INTEGRATION`. Though if `ENABLE_CUDA` is `ON`, then the OpenGL functionality should be turned on by default too.

**Comment by cenit** [comment]
> @cenit (@1027663760 too) this _should_ work on Windows now, hopefully...
> 
> I sat down with my coworker (thank you @skywhale!) and got it building on a windows machine.
> 
> I also tried to get the functionality behind a more proper feature flag, let me know if I messed that up.

thanks a lot

**Comment by cenit** [comment]
> @1027663760 it's a CMake option, `ENABLE_CUDA_OPENGL_INTEGRATION`. Though if `ENABLE_CUDA` is `ON`, then the OpenGL functionality should be turned on by default too.

you could make it a cmake_dependent_option, maybe?

**Comment by cenit** [comment]
and also please do a merge with master in order to verify if the CI errors are fixed or not, thanks

File: include/darknet.h (hunk 1 above)
Comment (cenit) [reply include/darknet.h:1036]: isn't CUDA_OPENGL_INTEGRATION defined only if GPU is defined too?

File: src/network_kernels.cu (hunk 3 above)
Comment (cenit) [reply src/network_kernels.cu:835]: it's ok, there is a lot of code copy-pasted here, so it's like an accepted method!

**Comment by cenit** [comment]
bump also the "version-date": "2021-04-16" that you can find in the vcpkg.json file, thanks

**Comment by bschwind** [comment]
@cenit I rebased on `master` - the version-date string is already set to "2021-04-16" in the latest version.

File: include/darknet.h (hunk 1 above)
Comment (bschwind) [reply include/darknet.h:1036]: Yep, I think this comment was left before I added the `CUDA_OPENGL_INTEGRATION` ifdef, so it should be covered now.

**Comment by cenit** [comment]
yes please bump it to the date of your latest commit, i will do it anyway. I don't bump it when i touch scripts, but it's better to update it when we modify sources

**Comment by bschwind** [comment]
Ah sorry, I misinterpreted your comment! I'll set it to today's date.

File: include/darknet.h [hunk 5]
```diff
@@ -1032,6 +1032,10 @@ LIB_API void diounms_sort(detection *dets, int total, int classes, float thresh,
 // network.h
 LIB_API float *network_predict(network net, float *input);
 LIB_API float *network_predict_ptr(network *net, float *input);
+#ifdef CUDA_OPENGL_INTEGRATION
```
Comment (bschwind) [inline include/darknet.h:1035]: I think I'm going to remove this `ifdef`, it doesn't seem harmful to leave the function definition in, and it's causing some issues when I try to generating Rust bindings to this code.

**Comment by bschwind** [comment]
I think this _should_ be ready for another review round. There's a failed CI check having to do with the CircleCI Pipeline, I'm not sure what that's about.

**Comment by cenit** [comment]
CircleCI pipeline is under @AlexeyAB control, not mine.
For me your work is ok apart from the following comment, but for merge let's wait for Alexey approval
My remaining question is this one:
did you try a fresh install with scripts/setup.sh and OpenGL integration on linux? Is it working or do we need to add libgles2-mesa-dev also to that script, as you did to GA pipelines?

**Comment by bschwind** [comment]
> did you try a fresh install with scripts/setup.sh and OpenGL integration on linux? Is it working or do we need to add libgles2-mesa-dev also to that script, as you did to GA pipelines?

@cenit Good question, I will try that out and report back here.

**Comment by bschwind** [comment]
> Is it working or do we need to add libgles2-mesa-dev also to that script, as you did to GA pipelines?

I honestly can't say if it's working or not. I just tried on a fresh Pop_OS! 22.04 and ran into various issues, but none were related to OpenGL specifically, so I guess I never made it far enough to see if it works.

I would guess that we do need to add it, assuming one of the other dependencies didn't already pull it in. I have no idea how to effectively modify these setup files correctly though, so if you could direct me on where to add it I can include it in this PR.

**Comment by cenit** [comment]
can you share some logs or a dockerfile with issues you encountered?
setup.sh should bring a vanilla (ubuntu-like) image ready to build darknet full-fledged. It is tested in CI, it was expecting it just required adding the libgl-mesa-dev to it

**Comment by bschwind** [comment]
@cenit sorry for the long delay. I don't have logs from when I first attempted it. My coworker and I are currently trying to get this running in a minimal dockerfile, if we're successful we'll post it here.

But speaking of `setup.sh`, it seems `libgles2-mesa-dev` is already included in the list of things to install:

https://github.com/AlexeyAB/darknet/blob/0faed3e60e52f742bbef43b83f6be51dd30f373e/scripts/setup.sh#L80

**Comment by skywhale** [comment]
@cenit Thank you so much for your help trying to get this merged. I'm working with @bschwind and was looking into the build errors he was seeing on a vanilla ubuntu-like environment. I'm also seeing some build errors, but before looking into it further, I thought I would ask if you have a minimal Dockerfile to reproduce a build environment for darknet. I tried to bootstrap my own, but I'm hitting some cmake configuration errors. This is the full output of the `build.ps1` script.

<details>
<summary>Output of the `build.ps1` script</summary>

```
Transcript started, output file is /darknet/build.log
Build script version 2.3.3, utils module version 0.0.4
PowerShell version:
Major  Minor  Patch  PreReleaseLabel BuildLabel
-----  -----  -----  --------------- ----------
7      3      1
vcpkg bootstrap script: bootstrap-vcpkg.sh
CUDA is enabled
CUDNN is enabled
OPENCV is enabled
OPENCV with CUDA extension is not enabled, you can enable it passing -EnableOPENCV_CUDA
VCPKG is enabled
VCPKG will be updated to latest version if found
VisualStudio integration is disabled
Yolo C# wrapper integration is disabled, please pass -EnableCSharpWrapper to the script to enable. You must be on Windows!
Ninja is enabled, please pass -DoNotUseNinja to the script to disable
ForceCPP build mode is disabled, please pass -ForceCPP to the script to enable
Using git from /usr/bin/git
This tool has been cloned with git and supports self-updating mechanism
This tool will not self-update sources
Using CMake from /usr/local/bin/cmake
cmake version 3.25.1

CMake suite maintained and supported by Kitware (kitware.com/cmake).
Using Ninja from /usr/bin/ninja
Ninja version 1.10.0
Found vcpkg in /darknet/vcpkg: /darknet/vcpkg
Already up to date.
Downloading vcpkg-glibc...
Setting up environment to use CMake generator: Ninja
Unable to find CUDA, if necessary please install it or define a CUDA_PATH env variable pointing to the install folder
Removing folder ./build_release
Configuring CMake project
CMake args: -G "Ninja"  -DCMAKE_BUILD_TYPE=Release -DENABLE_VCPKG_INTEGRATION:BOOL=ON -DX_VCPKG_APPLOCAL_DEPS_INSTALL=ON -DVCPKG_BUILD_OPENCV_WITH_CUDA:BOOL=OFF -DENABLE_DEPLOY_CUSTOM_CMAKE_MODULES:BOOL=ON -S ..
-- Setting default vcpkg target triplet to x64-linux-release
-- VCPKG found: /darknet/vcpkg
-- Using VCPKG integration
-- VCPKG_MANIFEST_FEATURES: cuda;cudnn;opencv-base
-- Running vcpkg install
Detecting compiler hash for triplet x64-linux-release...
The following packages will be built and installed:
  * at-spi2-atk[core]:x64-linux-release -> 2.38.0
  * at-spi2-core[core]:x64-linux-release -> 2.44.1
  * atk[core]:x64-linux-release -> 2.38.0#4
  * brotli[core]:x64-linux-release -> 1.0.9#4
  * bzip2[core,tool]:x64-linux-release -> 1.0.8#3
  * cairo[core,fontconfig,freetype,gobject,x11]:x64-linux-release -> 1.17.6#4
    cuda[core]:x64-linux-release -> 10.1#11
    cudnn[core]:x64-linux-release -> 7.6.5#8
  * curl[core,non-http,openssl,ssl]:x64-linux-release -> 7.86.0#1
  * dirent[core]:x64-linux-release -> 1.23.2#1
  * egl-registry[core]:x64-linux-release -> 2022-09-20
  * expat[core]:x64-linux-release -> 2.5.0#3
  * ffmpeg[avcodec,avdevice,avfilter,avformat,core,swresample,swscale]:x64-linux-release -> 5.0.2#1
  * fontconfig[core]:x64-linux-release -> 2.14.1#1
  * freetype[brotli,bzip2,core,png,zlib]:x64-linux-release -> 2.12.1#3
  * fribidi[core]:x64-linux-release -> 1.0.12
  * gdk-pixbuf[core]:x64-linux-release -> 2.42.9#4
  * getopt[core]:x64-linux-release -> 0#2
  * gettext[core,tools]:x64-linux-release -> 0.21#9
  * giflib[core]:x64-linux-release -> 5.2.1#3
  * glib[core]:x64-linux-release -> 2.75.1
  * gperf[core]:x64-linux-release -> 3.1#4
  * gtk3[core]:x64-linux-release -> 3.24.34#1
  * harfbuzz[core]:x64-linux-release -> 5.0.1#3
  * hdf5[core,szip,zlib]:x64-linux-release -> 1.12.2#4
  * leptonica[core]:x64-linux-release -> 1.82.0
  * libarchive[bzip2,core,libxml2,lz4,lzma,openssl,zstd]:x64-linux-release -> 3.6.1#1
  * libepoxy[core]:x64-linux-release -> 1.5.10#1
  * libffi[core]:x64-linux-release -> 3.4.4
  * libiconv[core]:x64-linux-release -> 1.17
  * libjpeg-turbo[core]:x64-linux-release -> 2.1.4
  * liblzma[core]:x64-linux-release -> 5.4.0
  * libpng[core]:x64-linux-release -> 1.6.39
  * libuuid[core]:x64-linux-release -> 1.0.3#11
  * libwebp[core,libwebpmux,nearlossless,simd]:x64-linux-release -> 1.2.4#1
  * libxml2[core]:x64-linux-release -> 2.10.3
  * lz4[core]:x64-linux-release -> 1.9.4#1
  * lzo[core]:x64-linux-release -> 2.10#7
    opencv[contrib,core,default-features,dnn,ffmpeg]:x64-linux-release -> 4.6.0
  * opencv4[contrib,core,default-features,dnn,ffmpeg,gtk,jpeg,png,quirc,tiff,webp]:x64-linux-release -> 4.6.0#10
  * openjpeg[core]:x64-linux-release -> 2.5.0
  * openssl[core]:x64-linux-release -> 3.0.7#2
  * pango[core]:x64-linux-release -> 1.50.12#1
  * pcre2[core]:x64-linux-release -> 10.40#1
  * pixman[core]:x64-linux-release -> 0.40.0#5
  * pkgconf[core]:x64-linux-release -> 1.8.0#3
  * protobuf[core]:x64-linux-release -> 3.21.8
  * pthread[core]:x64-linux-release -> 3.0.0#1
    pthreads[core]:x64-linux-release -> 3.0.0#13
  * quirc[core]:x64-linux-release -> 1.1#3
    stb[core]:x64-linux-release -> 2021-09-10#1
  * szip[core]:x64-linux-release -> 2.1.1#9
  * tesseract[core]:x64-linux-release -> 5.2.0#3
  * tiff[core,jpeg,lzma,zip]:x64-linux-release -> 4.5.0#1
  * vcpkg-cmake[core]:x64-linux-release -> 2022-11-13
  * vcpkg-cmake-config[core]:x64-linux-release -> 2022-02-06#1
  * vcpkg-cmake-get-vars[core]:x64-linux-release -> 2022-12-16
  * vcpkg-get-python-packages[core]:x64-linux-release -> 2022-06-30
  * vcpkg-pkgconfig-get-modules[core]:x64-linux-release -> 2022-02-10#1
  * vcpkg-tool-meson[core]:x64-linux-release -> 0.63
  * zlib[core]:x64-linux-release -> 1.2.13
  * zstd[core]:x64-linux-release -> 1.5.2#2
Additional packages (*) will be modified to complete this operation.
Restored 22 package(s) from /root/.cache/vcpkg/archives in 314.7 ms. Use --debug to see more details.
Installing 1/62 vcpkg-cmake-config:x64-linux-release...
Elapsed time to handle vcpkg-cmake-config:x64-linux-release: 1.247 ms
Installing 2/62 pthreads:x64-linux-release...
Elapsed time to handle pthreads:x64-linux-release: 376.3 us
Installing 3/62 stb:x64-linux-release...
Elapsed time to handle stb:x64-linux-release: 2.959 ms
Installing 4/62 cuda:x64-linux-release...
Elapsed time to handle cuda:x64-linux-release: 405.2 us
Installing 5/62 cudnn:x64-linux-release...
Elapsed time to handle cudnn:x64-linux-release: 557.7 us
Installing 6/62 vcpkg-cmake:x64-linux-release...
Elapsed time to handle vcpkg-cmake:x64-linux-release: 631.5 us
Installing 7/62 libwebp:x64-linux-release...
Elapsed time to handle libwebp:x64-linux-release: 3.471 ms
Installing 8/62 zlib:x64-linux-release...
Elapsed time to handle zlib:x64-linux-release: 1.054 ms
Installing 9/62 liblzma:x64-linux-release...
Elapsed time to handle liblzma:x64-linux-release: 2.404 ms
Installing 10/62 libjpeg-turbo:x64-linux-release...
Elapsed time to handle libjpeg-turbo:x64-linux-release: 3.327 ms
Installing 11/62 tiff:x64-linux-release...
Elapsed time to handle tiff:x64-linux-release: 2.075 ms
Installing 12/62 quirc:x64-linux-release...
Elapsed time to handle quirc:x64-linux-release: 898.5 us
Installing 13/62 libpng:x64-linux-release...
Elapsed time to handle libpng:x64-linux-release: 2.046 ms
Installing 14/62 vcpkg-tool-meson:x64-linux-release...
Elapsed time to handle vcpkg-tool-meson:x64-linux-release: 49.67 ms
Installing 15/62 bzip2:x64-linux-release...
Elapsed time to handle bzip2:x64-linux-release: 1.735 ms
Installing 16/62 brotli:x64-linux-release...
Elapsed time to handle brotli:x64-linux-release: 3.305 ms
Installing 17/62 freetype:x64-linux-release...
Elapsed time to handle freetype:x64-linux-release: 6.534 ms
Installing 18/62 harfbuzz:x64-linux-release...
Elapsed time to handle harfbuzz:x64-linux-release: 7.227 ms
Installing 19/62 pcre2:x64-linux-release...
Elapsed time to handle pcre2:x64-linux-release: 3.934 ms
Installing 20/62 libiconv:x64-linux-release...
Elapsed time to handle libiconv:x64-linux-release: 989.3 us
Installing 21/62 libffi:x64-linux-release...
Elapsed time to handle libffi:x64-linux-release: 1.591 ms
Installing 22/62 gettext:x64-linux-release...
Building gettext[core,tools]:x64-linux-release...
warning: -- Using community triplet x64-linux-release. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: /darknet/vcpkg/triplets/community/x64-linux-release.cmake
-- Using cached gettext-0.21.tar.gz.
-- Cleaning sources at /darknet/vcpkg/buildtrees/gettext/src/0.21-66fbad380c.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /darknet/vcpkg/downloads/gettext-0.21.tar.gz
-- Applying patch 0002-Fix-uwp-build.patch
-- Applying patch 0003-Fix-win-unicode-paths.patch
-- Applying patch 0004-Fix-uwp-tools-build.patch
-- Applying patch rel_path.patch
-- Applying patch android.patch
-- Applying patch gettext-tools_woe32dll_gettextsrc-exports.c.patch
-- Using source at /darknet/vcpkg/buildtrees/gettext/src/0.21-66fbad380c.clean
-- Getting CMake variables for x64-linux-release-rel
-- Configuring x64-linux-release-rel
-- Building x64-linux-release-rel
-- Installing x64-linux-release-rel
-- Performing post-build validation
Stored binary cache: "/root/.cache/vcpkg/archives/a4/a4042b848a2b460a2cdd27a20d58a7720a90e7e087ecfdbb4dcfb1f9a586ae64.zip"
Elapsed time to handle gettext:x64-linux-release: 2.42 min
Installing 23/62 dirent:x64-linux-release...
Elapsed time to handle dirent:x64-linux-release: 1.068 ms
Installing 24/62 glib:x64-linux-release...
Building glib[core]:x64-linux-release...
warning: -- Using community triplet x64-linux-release. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: /darknet/vcpkg/triplets/community/x64-linux-release.cmake
-- Using cached glib-2.75.1.tar.xz.
-- Cleaning sources at /darknet/vcpkg/buildtrees/glib/src/glib-2-cc73e48deb.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /darknet/vcpkg/downloads/glib-2.75.1.tar.xz
-- Applying patch use-libiconv-on-windows.patch
-- Applying patch libintl.patch
-- Using source at /darknet/vcpkg/buildtrees/glib/src/glib-2-cc73e48deb.clean
-- Getting CMake variables for x64-linux-release-rel
-- Configuring x64-linux-release-rel
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: /usr/bin/python3 /darknet/build_release/vcpkg_installed/x64-linux-release/tools/meson/meson.py -Dselinux=disabled -Dlibmount=disabled -Dgtk_doc=false -Dinstalled_tests=false -Dlibelf=disabled -Dman=false -Dtests=false -Dxattr=false --buildtype plain --backend ninja --wrap-mode nodownload --libdir lib --native /darknet/vcpkg/buildtrees/glib/meson-x64-linux-release-rel.log -Ddebug=false --prefix /darknet/vcpkg/packages/glib_x64-linux-release -Dcmake_prefix_path=['/darknet/build_release/vcpkg_installed/x64-linux-release','/darknet/build_release/vcpkg_installed/x64-linux-release/debug'] /darknet/vcpkg/buildtrees/glib/src/glib-2-cc73e48deb.clean
    Working Directory: /darknet/vcpkg/buildtrees/glib/x64-linux-release-rel
    Error code: 1
    See logs for more information:
      /darknet/vcpkg/buildtrees/glib/config-x64-linux-release-rel-meson-log.txt.log
      /darknet/vcpkg/buildtrees/glib/config-x64-linux-release-rel-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_meson.cmake:432 (vcpkg_execute_required_process)
  ports/glib/portfile.cmake:33 (vcpkg_configure_meson)
  scripts/ports.cmake:147 (include)


error: building glib:x64-linux-release failed with: BUILD_FAILED
error: Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
    https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+glib
You can submit a new issue at:
    https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[glib]+Build+error
Include '[glib] Build error' in your bug report title, the following version information in your bug description, and attach any relevant failure logs from above.
    vcpkg-tool version: 2022-12-14-7ae0d8527fb488fde10a89c2813802dc9b03b6f9
    vcpkg-scripts version: c30de8e13 2023-01-10 (5 hours ago)


Please use the prefilled template from /darknet/build_release/vcpkg_installed/vcpkg/issue_body.md when reporting your issue.
-- Running vcpkg install - failed
-- Configuring incomplete, errors occurred!
Config failed! Exited with error code 1.
Exception: ScriptHalted
```
</details>

And my bootstrapping Dockerfile:

```docker
FROM ubuntu:20.04

RUN apt-get update
RUN apt-get install -y lsb-release git curl zip mono-devel wget libssl-dev

# cmake >3.18 is required
RUN curl -O https://github.com/Kitware/CMake/releases/download/v3.25.1/cmake-3.25.1-linux-x86_64.sh
RUN mkdir /usr/local/cmake
RUN sh cmake-*.sh --skip-license --prefix=/usr/local/cmake
RUN ln -s /usr/local/cmake/bin/cmake /usr/local/bin

# Mock sudo command
RUN echo "#!/bin/sh\n$@" > /usr/local/bin/sudo
RUN chmod +x /usr/local/bin/sudo

WORKDIR /darknet

# Needed to run this twice
CMD ["./scripts/setup.sh", "-InstallTOOLS", "-InstallCUDA", "-BypassDRIVER"]
```

**Comment by cenit** [comment]
can you please merge with master?

**Comment by bschwind** [comment]
Hi @cenit , I can still rebase on `master` if you like, this was just bad timing coinciding with longer travel.

**Comment by cenit** [comment]
i merged manually your commits into master :)

**Pull request closed without merging**